numpy/lib/arraysetops.py

```diff
@@ -324,7 +325,15 @@ def _unique1d(ar, return_index=False, return_inverse=False,
         aux = ar
     mask = np.empty(aux.shape, dtype=np.bool_)
     mask[:1] = True
-    mask[1:] = aux[1:] != aux[:-1]
+    if aux.shape[0] > 0 and aux.dtype.kind in "cfmM" and np.isnan(aux[-1]):
+        # Ensure that `NaT` is used for time-like dtypes
+        nan = np.array(np.nan).astype(aux.dtype)
+        aux_firstnan = np.searchsorted(aux, nan, side='left')
+        mask[1:aux_firstnan] = (aux[1:aux_firstnan] != aux[:aux_firstnan - 1])
+        mask[aux_firstnan] = True
+        mask[aux_firstnan + 1:] = False
```

Comment:
This `aux[-1]` is really smart. I will look into complex now, but I feel it becomes harder now for a novice like me.

Reply:
I would also agree to consider all complex NaNs as equivalent, although I can imagine that others might have a different view. If we agree on this, then we need to decide, which representant from all the input NaNs to put into the return array. In my proposal, it is the first element `v` with true value of `np.isnan(v)` found in the sorted array `aux`. The documentation and the release note is updated accordingly.

Reply:
So clearly `isnan` and `searchsorted` do not care whether to real or imaginary part is `nan`.
@seberg do you know if there are any functions within numpy where this distinction _does_ matter?
If not, I'd say that it would be consistent to treat all complex NaNs as equivalent.

Reply:
@BvB93 I am not sure if the distinction matters anywhere, probably not (aside from things like `.real`, since it is a view that would not ensure a NaN real part when the imaginary part is NaN).  I would not trust this to be tested for anywhere, but I guess most functionality either correctly inherits behaviour from glibc or is naively correct (that is with the probable exception of warning flags.)

I will not dig it up right now, but can do so some time. There was a discussion semi recently (probably to do with warning flags), and if I remember right, all complex NaNs are considered identical by an IEEE standard. (I don't remember if it said anything about infinities.)

Reply:
Just to verify: it seems we all agree that handling of `nan`, as is done in this PR, is desirable?

Reply:
Yes, I am happy with it, right now it is not consistent/correct for complex though.  Maybe we can live with that, I have not checked how complex it would be to fix the complex search-sorted.